### PR TITLE
Support executing tasks with keywords

### PR DESCRIPTION
**Pull request by omus**
- [x] Requires #47.

### CODE REVIEW

File: Ray.jl/src/runtime.jl [hunk 1]
```diff
@@ -181,16 +175,21 @@ function task_executor(ray_function, returns_ptr, task_args_ptr, task_name,
                                 rayjll.unwrap_function_descriptor(fd),
                                 get_current_job_id())
         # TODO: write generic Ray.put and Ray.get functions and abstract over this buffer stuff
-        args = map(task_args) do arg
+        flattened = map(task_args) do arg
```
Comment (glennmoy) [inline Ray.jl/src/runtime.jl:178]: could we map over `Ray.get` here?

File: Ray.jl/src/remote_function.jl [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Underscores are not allowed to be used as Julia function keywords
+# https://docs.julialang.org/en/v1/manual/variables/#man-allowed-variable-names
+const ARG_KEY = :_
+
+function flatten_args(args, kwargs)
```
Comment (glennmoy) [inline Ray.jl/src/remote_function.jl:5]: maybe write a cursory unit test for this?

File: Ray.jl/src/runtime.jl [hunk 3]
```diff
@@ -181,16 +175,21 @@ function task_executor(ray_function, returns_ptr, task_args_ptr, task_name,
...
         end
+        args, kwargs = recover_args(flattened)
 
-        arg_string = join(string.("::", typeof.(args)), ", ")
-        @info "Calling $func($arg_string)"
+        @info begin
+            args_str = join([string("::", typeof(arg)) for arg in args], ", ")
+            kwargs_str = join([string(k, "::", typeof(v)) for (k, v) in kwargs], ", ")
+            args_kwargs_str = join([args_str, kwargs_str], "; ")
+            "Calling $func($args_kwargs_str)"
```
Comment (kleinschmidt) [inline Ray.jl/src/runtime.jl:192]: this seems more natural to me:
```suggestion
            "Calling $func($args_str; $kwargs_str)"
```

File: Ray.jl/src/runtime.jl [hunk 4]
```diff
@@ -181,16 +175,21 @@ function task_executor(ray_function, returns_ptr, task_args_ptr, task_name,
...
             io = IOBuffer(v)
             return deserialize(io)
         end
+        args, kwargs = recover_args(flattened)
 
-        arg_string = join(string.("::", typeof.(args)), ", ")
-        @info "Calling $func($arg_string)"
+        @info begin
+            args_str = join([string("::", typeof(arg)) for arg in args], ", ")
+            kwargs_str = join([string(k, "::", typeof(v)) for (k, v) in kwargs], ", ")
```
Comment (kleinschmidt) [inline Ray.jl/src/runtime.jl:190]: I liked the broadcasting but maybe it's too clever?  or doesn't work with the kv pairs I guess...  if you're gonna use generators might as well make them fully lazy to not allocate teh intermediate arrays

```suggestion
            args_str = join((string("::", typeof(arg)) for arg in args), ", ")
            kwargs_str = join((string(k, "::", typeof(v)) for (k, v) in kwargs), ", ")
```

**Review by kleinschmidt — APPROVED** [review]
only extremely nitpicky comments! otherwise looks good

File: Ray.jl/src/runtime.jl (hunk 3 above)
Comment (omus) [reply Ray.jl/src/runtime.jl:192]: I agree with you. I meant to get back to this as what I was trying to do was avoid having the "; " present when no keywords were used. I'll update the logic here to be:
```julia
isempty(kwargs_str) || (args_str *= "; " * kwargs_str)
"Calling $func($args_str)"
```

Comment (kleinschmidt) [reply Ray.jl/src/runtime.jl:192]: could make `kwargs_str = isempty(kwargs) ? "" : string("; ", ...)`

Comment (omus) [reply Ray.jl/src/runtime.jl:192]: Unfortunately with that we end up with `$func($args_str$kwargs_str)`

Comment (omus) [reply Ray.jl/src/runtime.jl]: Ended up with this:
```julia
param_str = join((string("::", typeof(arg)) for arg in args), ", ")
if !isempty(kwargs)
    param_str *= "; "
    param_str *= join((string(k, "::", typeof(v)) for (k, v) in kwargs), ", ")
end
"Calling $func($param_str)"
```
Having `param_str *= "; "` on a separate line just avoids hitting the line length limit

**Pull request merged**